Support reading charmcraft.yaml in testing harness

syu-w opened the pull request:
Try read unified charmcraft.yaml first, that may included metadata, actions, and config. Ignore metadata.yaml, actions.yaml, config.yaml if they are exists in charmcraft.yaml

Fixes https://github.com/canonical/operator/issues/969

benhoyt commented on ops/main.py:
```diff
@@ -384,12 +387,28 @@ def main(charm_class: Type[ops.charm.CharmBase],
+        actions_metadata = charmcraft_metadata.get("actions")
+
+    if not metadata:
+        metadata = (charm_dir / "metadata.yaml").read_text()
```
In charmcraft's behaviour, is there any merging? Does charmcraft raise an error if the charm has metadata in both `charmcraft.yaml` and `metadata.yaml`? Should we raise an error too, to catch it early?

syu-w replied:
charmcraft will raise error when it found metadata in `charmcraft.yaml` but `metadata.yaml` also exists. `actions` and `config` also have the same behavior.

benhoyt commented on ops/main.py:
```diff
@@ -384,12 +387,28 @@ def main(charm_class: Type[ops.charm.CharmBase],
+        charmcraft_yaml = charmcraft_meta.read_text()
+        charmcraft_metadata: dict[str, Any] = yaml.safe_load(charmcraft_yaml)
+        if any(
+            meta_key in charmcraft_metadata for meta_key in ("name", "summary", "description")
```
Is this the behaviour of / test used by charmcraft too? Can you please point to the code there?

benhoyt commented on ops/testing.py:
```diff
@@ -394,9 +394,30 @@ def _create_meta(self, charm_metadata: Optional[YAMLStringOrFile],
         """Create a CharmMeta object.
 
         Handle the cases where a user doesn't supply explicit metadata snippets.
+        Will try to load metadata/config from charm_dir/charmcraft.yaml first, then
+        charm_dir/metadata.yaml and/or charm_dir/actions.yaml if charmcraft.yaml
+        not include metadata and/or actions.
```
It looks like this method doesn't deal with config. Also a few minor tweaks for English and clarity. Is this correct?

```suggestion
        This will try to load metadata from ``<charm_dir>/charmcraft.yaml`` first, then
        ``<charm_dir>/metadata.yaml`` if charmcraft.yaml does not include metadata,
        and ``<charm_dir>/actions.yaml`` if charmcraft.yaml does not include actions.
```

benhoyt commented on ops/testing.py:
```diff
@@ -394,9 +394,30 @@ def _create_meta(self, charm_metadata: Optional[YAMLStringOrFile],
+        # Should fall back to metadata.yaml if both are not provided
+        if charm_metadata is None and charmcraft_metadata:
+            if any(
+                meta_key in charmcraft_metadata for meta_key in ("name", "summary", "description")
```
A lot of this logic looks very similar to the logic in `main.py`. Ideally we can use `CharmMeta.from_files` or similar once that exists.

benhoyt replied:
We also need a couple of tests of this behaviour (for metadata and for config) in `test/test_testing.py`.

syu-w replied:
Added tests.

benhoyt reviewed (changes requested):
Thanks for this PR! I've added a few comments/questions.

syu-w commented on ops/charm.py:
```diff
@@ -1118,18 +1118,22 @@ def __init__(self, raw: Optional[Dict[str, Any]] = None,
     def from_yaml(
             cls, metadata: Union[str, TextIO],
-            actions: Optional[Union[str, TextIO]] = None) -> 'CharmMeta':
+            actions: Optional[Union[str, TextIO, Dict[str, Any]]] = None) -> 'CharmMeta':
```
Added `from_dict`

syu-w commented on ops/main.py:
```diff
@@ -384,12 +387,28 @@ def main(charm_class: Type[ops.charm.CharmBase],
-        actions_metadata = actions_meta.read_text()
-    else:
-        actions_metadata = None
+    # Check charmcraft.yaml first and see if it has the metadata data in it.
```
OK, I removed this part from `main.py` that seems not used when call testing.

syu-w commented:
I removed the change in `main.py` that doesn't look like it's currently being used. `ops` in charm will still have separate yaml files as for now. Improved import logic, added `from_dict` and some tests.

benhoyt commented on ops/testing.py:
```diff
@@ -389,57 +389,104 @@ def cleanup(self) -> None:
         charm_dir = pathlib.Path(filename).parents[1]
 
+        charm_metadata: Optional[Dict[str, Any]] = None
+        charmcraft_metadata: Optional[dict[str, Any]] = None
```
I'm not actually sure why this passes tests on Python 3.8, but the `dict[K, V]` syntax with lowercase `dict` type was only added in Python 3.9 (see https://docs.python.org/3/whatsnew/3.9.html and PEP 585). It would be interesting to figure out why the tests don't fail on 3.8! When I do this locally I get this error:

```
$ python3.8
Python 3.8.17 (default, Jun  8 2023, 17:21:52) 
[GCC 11.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> d: dict[str, int] = {}
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'type' object is not subscriptable
```

In any case, let's use `Dict` here for consistency. Also below.

benhoyt commented on ops/testing.py:
```diff
@@ -389,57 +389,104 @@ def cleanup(self) -> None:
+        else:
+            # Check charmcraft.yaml for metadata if no metadata is provided
+            if charmcraft_metadata is not None:
+                if any(
+                    meta_key in charmcraft_metadata
+                    for meta_key in ("name", "summary", "description")
+                ):
```
Style nit, but arguably a bit easier to read on two lines than four (and this is all about "meta", so no need for the `meta_` prefix in the loop variable name):


```suggestion
                meta_keys = ["name", "summary", "description"]
                if any(key in charmcraft_metadata for key in meta_keys):
```

benhoyt commented on ops/testing.py:
```diff
@@ -389,57 +389,104 @@ def cleanup(self) -> None:
+                    action_metadata = yaml.safe_load(actions_path.read_text())
+                    self._charm_dir = charm_dir
+
+        return CharmMeta.from_dict(charm_metadata, action_metadata)
```
This is where (unless I'm missing something) we can just use `CharmMeta(charm_metadata, action_metadata)`.

benhoyt commented on test/test_testing.py:
```diff
@@ -1262,6 +1262,32 @@ def test_metadata_from_directory(self):
 
+    def test_metadata_from_directory_charmcraft_yaml(self):
+        tmp = pathlib.Path(tempfile.mkdtemp())
+        self.addCleanup(shutil.rmtree, str(tmp))
```
Nit: file operations like `rmtree` work fine with `pathlib.Path` objects, so no need for the `str()` here.

benhoyt commented on test/test_testing.py:
```diff
@@ -1262,6 +1262,32 @@ def test_metadata_from_directory(self):
+        self.addCleanup(shutil.rmtree, str(tmp))
+        charmcraft_filename = tmp / 'charmcraft.yaml'
+        with charmcraft_filename.open('wt') as charmcraft_yaml:
+            charmcraft_yaml.write(textwrap.dedent('''
```
Simpler to use `charmcraft_filename.write_text()` rather than `with ... write`?

benhoyt commented on test/test_testing.py:
```diff
@@ -1302,6 +1328,56 @@ def test_config_from_directory(self):
+        self.addCleanup(shutil.rmtree, str(tmp))
+        charmcraft_filename = tmp / 'charmcraft.yaml'
+        with charmcraft_filename.open('wt') as charmcraft_yaml:
+            charmcraft_yaml.write(textwrap.dedent('''
```
Similar nits here about not need `str()` and using `write_text`.

benhoyt commented on test/test_testing.py:
```diff
@@ -1302,6 +1328,56 @@ def test_config_from_directory(self):
+
+            config:
+                options:
+                    opt_str:
```
We don't need so many options to test this -- we're just testing that it gets loaded from the right place. I think two would be fine, maybe `opt_str` and `opt_int`.

benhoyt reviewed (commented):
This is looking very close, thanks. A few nit/style comments and one comment that I don't think `from_dict` is actually needed (just use the initialiser).

syu-w commented:
Thanks for the comments, updated.
